File: 天梯赛/L3-002. 堆栈.cpp
#include <iostream>
#include <cstring>
#include <stack> 
#include <string>
#include <algorithm>
using namespace std;
typedef long long ll;
const int inf=0x3f3f3f3f;
const int maxn=1e5+5;
#define lowbit(i) ((i)&(-i))
int c[maxn];
stack<int> s;
void update(int x,int v){
	for(;x<maxn;x+=lowbit(x))c[x]+=v;
}
int getsum(int x){
	int res=0;
	for(;x>=1;x-=lowbit(x))res+=c[x];
	return res;
}
int getmedium(){
	int l=1,r=maxn,k=(s.size()+1)>>1;
	while(l<r){
		int mid=(l+r)>>1;
		if(getsum(mid)>=k)r=mid;
		else l=mid+1;
	}
	return l;
}
int main(){
	ios::sync_with_stdio(false);
	string ss;
	int n;
	cin>>n;
	while(n--){
		cin>>ss;
		if(ss[1]=='o'){
			if(s.empty())cout<<"Invalid\n";
			else{
				update(s.top(),-1);
				cout<<s.top()<<"\n";
				s.pop();
			}
		}else if(ss[1]=='e'){
			if(s.empty())cout<<"Invalid\n";
			else cout<<getmedium()<<"\n";
		}else{//push
			int v;
			cin>>v;
			s.push(v);
			update(v,1);
		}
	}
	return 0;
}

File: 天梯赛/java/L3-002. 堆栈.java
import java.util.Arrays;
import java.util.LinkedList;
import java.util.List;
import java.util.Scanner;
import java.util.Stack;

public class Main {
	public static final int maxn=100001;
	public static int c[]=new int [maxn];
	public static Stack<Integer> s=new Stack<>();
	public static void main(String[] args) {
		Scanner in=new Scanner(System.in);
		int n=in.nextInt(),a=0;
		for(int i=0;i<n;++i) {
			String str=in.next();
			if(str.charAt(1)=='o') {
				if(s.empty())System.out.println("Invalid");
				else {
					update(s.peek(),-1);
					System.out.println(s.peek());
					s.pop();
				}
			}else if(str.charAt(1)=='e') {
				if(s.empty())System.out.println("Invalid");
				else PeekMedian();
			}else {
				a=in.nextInt();
				s.push(a);
				update(a,1);
			}
		}
	}

	public static int lowbit(int i) {
		return (i)&(-i);
	}
	
	public static void update(int x,int v) {
		for(;x<maxn;x+=lowbit(x))c[x]+=v;
	}
	
	public static int getsum(int x) {
		int res=0;
		for(;x>=1;x-=lowbit(x))res+=c[x];
		return res;
	}
	
	public static void PeekMedian() {
		int left=1,right=maxn,mid,k=(s.size()+1)/2;
		while(left<right) {
			mid=(left+right)>>1;
			if(getsum(mid)>=k)right=mid;
			else left=mid+1;
		}
		System.out.println(left);
	}
}

